lib/base_service.ts

```diff
@@ -129,12 +145,58 @@ export class BaseService {
     if (this._options.api_key) {
       _credentials.api_key = this._options.api_key;
     }
+    if (this._options.platform_api_key) {
+      _credentials.platform_api_key = this._options.platform_api_key;
+    }
     if (this._options.url) {
       _credentials.url = this._options.url;
     }
     return _credentials;
   }
 
+/**
+ * Wrapper around `sendRequest` that manages access tokens if applicable
+ * 1. If tokens are unsupported or not being used, call `sendRequest`
+ * 2. If tokens are being used but no token is saved, request an access token
+ * 3. If a token is saved but it is expired, refresh the access token
+ * 4. If token is saved and valid, use it to authenticate the service request
+ *
+ * @param {Object} parameters - service request options passed in by user
+ * @param {Function} _callback - callback function to pass the reponse back to
+ * @returns {ReadableStream|undefined}
+ */
+  protected createRequest(parameters, _callback) {
+    // condition 1: using basic or other auth, no need for any tokens
+    if (this.notUsingTokens()) {
+      return sendRequest(parameters, _callback);
+
+      // condition 2: no token has been retrieved, make a call to get the token
+    } else if (this.needToRequestToken()) {
+      tokenRequest(this._options.platform_api_key, (tokenResponse) => {
+        this.setTokenObject(tokenResponse);
+        parameters.defaultOptions.headers.Authorization =
+          `Bearer ${this.tokenInfo.access_token}`;
+        return sendRequest(parameters, _callback);
+      });
+
+      // condition 3: the token was retrieved, but has expired
+    } else if (this.tokenIsExpired()) {
+      tokenRequest(this.tokenInfo.refresh_token, (tokenResponse) => {
+        this.setTokenObject(tokenResponse);
+        parameters.defaultOptions.headers.Authorization =
+          `Bearer ${this.tokenInfo.access_token}`;
+        return sendRequest(parameters, _callback);
+        // this 'true' argument specifies that we're refreshing an existing token
+      }, true);
```

Comment:
I'm glad you have this comment. I was just wondering, "both of these calls to `tokenRequest` look the same, but one passes the API key and the other passes the refresh token...how does it know?"

It would be pretty easy to miss this, though. I wonder if it would make more sense to move it into the second argument position and pass it in line 175 as well?

`tokenRequest(this._options.iam_apikey, false, (tokenResponse) => { ... })`
`tokenRequest(this.tokenInfo.refresh_token, true, (tokenResponse) => { ... })`

Reply:
I've actually already changed this to use two different functions: one for requesting a token and one for refreshing. Even though the code is similar, I think it is way more clear. Would like to get other opinions though, once I push.